### PR TITLE
Support different app.go's for consumer and provider

### PR DESCRIPTION
**Pull request by danwt**
## Issues

Mostly completes

- https://github.com/cosmos/interchain-security/issues/58

Related to

- https://github.com/cosmos/interchain-security/issues/59
- https://github.com/cosmos/interchain-security/issues/77

## Overview

There is a goal to split provider/consumer apps so that people can build from a minimal viable consumer chain. This PR is paired with [another PR in informalsystems/ibc-go](https://github.com/informalsystems/ibc-go/pull/1). The two PR's together

1. Split the app.go into two app.go's. One for parent and child.
2. Modify ibc-go testing to be able to use two different app.go's.

Therefore this PR should be merged together with 

- https://github.com/informalsystems/ibc-go/pull/1

Note: there are two app.go's but they are _identical_.

### CODE REVIEW

**Comment by danwt** [comment]
There are at least a few things to do before this can be merged

1. Merge https://github.com/informalsystems/ibc-go/pull/1
2. Update go.mod to replace ibc-go with the updated ibc-go (after (1))

**Comment by danwt** [comment]
Hi @AdityaSripal and @sainoe. It would be great to get your eyes on this. The PR enables 2 app.go's. The changes look bigger than they are because the old app/ directory is duplicated -> app_consumer/, app_provider/ and all the calls to `app` were changed in the tests. Also, I slightly updated the integration tests to use two binaries.

**Review by alexanderbez — COMMENTED** [review]
Just a small nit. I'd have it as `app/consumer/app.go` and `app/producer/app.go`.

File: x/ccv/consumer/module_test.go [hunk 1]
```diff
@@ -48,9 +45,7 @@ type ConsumerTestSuite struct {
 }
 
 func (suite *ConsumerTestSuite) SetupTest() {
-	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 2)
-	suite.providerChain = suite.coordinator.GetChain(ibctesting.GetChainID(1))
-	suite.consumerChain = suite.coordinator.GetChain(ibctesting.GetChainID(2))
+	suite.coordinator, suite.providerChain, suite.consumerChain = simapp.NewProviderConsumerCoordinator(suite.T())
```
Comment (danwt) [inline x/ccv/consumer/module_test.go:48]: @sainoe I was wondering about this change in particular. There are three more similar instances (I will tag you).

I will also tag you in the `simapp.NewProviderConsumerCoordinator` code

File: x/ccv/consumer/keeper/keeper_test.go [hunk 2]
```diff
@@ -45,9 +42,7 @@ type KeeperTestSuite struct {
 }
 
 func (suite *KeeperTestSuite) SetupTest() {
-	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 2)
-	suite.providerChain = suite.coordinator.GetChain(ibctesting.GetChainID(1))
-	suite.consumerChain = suite.coordinator.GetChain(ibctesting.GetChainID(2))
+	suite.coordinator, suite.providerChain, suite.consumerChain = simapp.NewProviderConsumerCoordinator(suite.T())
```
Comment (danwt) [inline x/ccv/consumer/keeper/keeper_test.go:45]: @sainoe #2

File: x/ccv/provider/provider_test.go [hunk 3]
```diff
@@ -60,11 +56,9 @@ type ProviderTestSuite struct {
 }
 
 func (suite *ProviderTestSuite) SetupTest() {
-	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 1)
-	suite.providerChain = suite.coordinator.GetChain(ibctesting.GetChainID(1))
-	// create consumer chain with provider chain valset
-	suite.CreateConsumerChain()
-	suite.consumerChain = suite.coordinator.GetChain(ibctesting.GetChainID(2))
+
+	suite.coordinator, suite.providerChain, suite.consumerChain = simapp.NewProviderConsumerCoordinator(suite.T())
```
Comment (danwt) [inline x/ccv/provider/provider_test.go:60]: @sainoe #3

File: x/ccv/provider/keeper/keeper_test.go [hunk 4]
```diff
@@ -44,9 +41,7 @@ type KeeperTestSuite struct {
 }
 
 func (suite *KeeperTestSuite) SetupTest() {
-	suite.coordinator = ibctesting.NewCoordinator(suite.T(), 2)
-	suite.providerChain = suite.coordinator.GetChain(ibctesting.GetChainID(1))
-	suite.consumerChain = suite.coordinator.GetChain(ibctesting.GetChainID(2))
+	suite.coordinator, suite.providerChain, suite.consumerChain = simapp.NewProviderConsumerCoordinator(suite.T())
```
Comment (danwt) [inline x/ccv/provider/keeper/keeper_test.go:44]: @sainoe #4

File: testutil/simapp/simapp.go [hunk 5]
```diff
@@ -52,10 +37,42 @@ var defaultConsensusParams = &abci.ConsensusParams{
...
+func NewProviderConsumerCoordinator(t *testing.T) (*ibctesting.Coordinator, *ibctesting.TestChain, *ibctesting.TestChain) {
+	coordinator := NewBasicCoordinator(t)
+	chainID := ibctesting.GetChainID(0)
+	coordinator.Chains[chainID] = ibctesting.NewTestChain(t, coordinator, SetupTestingappProvider, chainID)
+	providerChain := coordinator.GetChain(chainID)
+	chainID = ibctesting.GetChainID(1)
+	coordinator.Chains[chainID] = ibctesting.NewTestChainWithValSet(t, coordinator,
+		SetupTestingappConsumer, chainID, providerChain.Vals, providerChain.Signers)
+	consumerChain := coordinator.GetChain(chainID)
+	return coordinator, providerChain, consumerChain
```
Comment (danwt) [inline testutil/simapp/simapp.go:77]: @sainoe This is the biggest change

Comment (danwt) [reply testutil/simapp/simapp.go:77]: Thanks by the way

File: testutil/simapp/simapp.go [hunk 6]
```diff
@@ -52,10 +37,42 @@ var defaultConsensusParams = &abci.ConsensusParams{
...
+func NewProviderConsumerCoordinator(t *testing.T) (*ibctesting.Coordinator, *ibctesting.TestChain, *ibctesting.TestChain) {
+	coordinator := NewBasicCoordinator(t)
+	chainID := ibctesting.GetChainID(1)
+	coordinator.Chains[chainID] = ibctesting.NewTestChain(t, coordinator, SetupTestingappProvider, chainID)
+	providerChain := coordinator.GetChain(chainID)
+	chainID = ibctesting.GetChainID(2)
+	coordinator.Chains[chainID] = ibctesting.NewTestChainWithValSet(t, coordinator,
+		SetupTestingAppConsumer, chainID, providerChain.Vals, providerChain.Signers)
+	consumerChain := coordinator.GetChain(chainID)
+	return coordinator, providerChain, consumerChain
```
Comment (danwt) [inline testutil/simapp/simapp.go:77]: @sainoe This is the change that might have impact. Thanks for reviewing.

**Review by sainoe — APPROVED** [review]
Great work @danwt!
Thanks for the test chains instantiation refactoring into one function.
LGTM

**Pull request merged**